packages/artifact/src/internal/download-http-client.ts

```diff
@@ -0,0 +1,195 @@
+import * as fs from 'fs'
+import * as zlib from 'zlib'
+import {
+  getArtifactUrl,
+  getRequestOptions,
+  isSuccessStatusCode,
+  isRetryableStatusCode,
+  createHttpClient
+} from './utils'
+import {URL} from 'url'
+import {ListArtifactsResponse, QueryArtifactResponse} from './contracts'
+import {IHttpClientResponse} from '@actions/http-client/interfaces'
+import {HttpManager} from './http-manager'
+import {DownloadItem} from './download-specification'
+import {
+  getDownloadFileConcurrency,
+  getRetryWaitTimeInMilliseconds
+} from './config-variables'
+import {warning} from '@actions/core'
+import {IncomingHttpHeaders} from 'http'
+
+export class DownloadHttpClient {
+  // http manager is used for concurrent connection when downloading mulitple files at once
+  private downloadHttpManager: HttpManager
+
+  constructor() {
+    this.downloadHttpManager = new HttpManager(getDownloadFileConcurrency())
+  }
+
+  /**
+   * Gets a list of all artifacts that are in a specific container
+   */
+  async listArtifacts(): Promise<ListArtifactsResponse> {
+    const artifactUrl = getArtifactUrl()
+    // use the first client from the httpManager, `keep-alive` is not used so the connection will close immediatly
+    const client = this.downloadHttpManager.getClient(0)
+    const requestOptions = getRequestOptions('application/json')
+
+    const rawResponse = await client.get(artifactUrl, requestOptions)
+    const body: string = await rawResponse.readBody()
+    if (isSuccessStatusCode(rawResponse.message.statusCode) && body) {
+      return JSON.parse(body)
+    }
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(`Unable to list artifacts for the run`)
+  }
+
+  /**
+   * Fetches a set of container items that describe the contents of an artifact
+   * @param artifactName the name of the artifact
+   * @param containerUrl the artifact container URL for the run
+   */
+  async getContainerItems(
+    artifactName: string,
+    containerUrl: string
+  ): Promise<QueryArtifactResponse> {
+    // the itemPath search parameter controls which containers will be returned
+    const resourceUrl = new URL(containerUrl)
+    resourceUrl.searchParams.append('itemPath', artifactName)
+
+    // no concurrent calls so a single httpClient without the http-manager is sufficient
+    const client = createHttpClient()
```

Comment:
Why not do `const client = this.downloadHttpManager.getClient(0)`?  We don't need to recreate the http client.

Reply:
Oops, missed this one. Got `listArtifacts` but accidentally skipped over this method. Nice catch, should be the same thing